virttest/qemu_devices/qcontainer.py

```diff
@@ -1720,6 +1720,147 @@ def images_define_by_params(self, name, image_params, media=None,
+        :param index: the index of the serial device (optional)
+        :return: the device list that construct the serial device
+        """
+        serial_type = serial_params['serial_type']
+        virtio_port_type = serial_params.get('virtio_port_type')
+        machine = serial_params.get('machine_type')
+        if virtio_port_type:
+            if "-mmio" in machine:
+                controller_suffix = 'device'
+            elif machine.startswith("s390"):
+                controller_suffix = 'ccw'
+            else:
+                controller_suffix = 'pci'
+            serial_type = '%s-%s' % (serial_type, controller_suffix)
+            virtio_port_type = "virt%s" % virtio_port_type
+
+        prefix = serial_params.get('virtio_port_name_prefix')
+        name = serial_id
+        name = prefix + name if prefix else name
+        serial_params['name'] = name
+
+        chardev_args = {'logfile': serial_params.get('chardev_logfile'),
+                        'logappend': serial_params.get('chardev_logappend')}
+        backend = serial_params.get('chardev_backend', 'unix_socket')
+        if backend not in ['socket', 'vc', 'ringbuf',
+                           'spiceport', 'spicevmc']:
+            chardev_args.update({'mux': serial_params.get('chardev_mux')})
+        if backend in ['file', 'pipe', 'serial',
+                       'tty', 'parallel', 'parport']:
+            chardev_args.update({'path': file_name})
+        elif backend in ['spicevmc', 'spiceport']:
+            chardev_args.update({'debug': serial_params.get('chardev_debug'),
+                                 'name': name})
+        elif backend in ['stdio']:
+            chardev_args.update({'signal': serial_params.get('chardev_signal')})
+        elif 'unix' in backend:  # unix_socket
+            chardev_args.update({'path': file_name})
+        elif 'tcp' in backend or backend in ['udp']:  # tcp_socket & udp
+            chardev_args.update({'host': serial_params.get('chardev_host'),
+                                 'port': serial_params.get('chardev_port'),
+                                 'ipv4': serial_params.get('chardev_ipv4'),
+                                 'ipv6': serial_params.get('chardev_ipv6')})
+            if backend in ['udp']:
+                chardev_args.update(
+                    {'localaddr': serial_params.get('chardev_localaddr'),
+                     'localport': serial_params.get('chardev_localport')})
+            else:
+                chardev_args.update({'to': serial_params.get('chardev_to'),
+                                     'nodelay': serial_params.get('chardev_nodelay')})
+        if 'socket' in backend:  # tcp_socket & unix_socket
+            chardev_args.update({'server': serial_params.get('chardev_server', 'yes'),
+                                 'nowait': serial_params.get('chardev_nowait', 'yes'),
+                                 'reconnect': serial_params.get('chardev_reconnect')})
```

Comment:
New design: param name is `chardev_$option_name`.
Why don't use class CharDevice directly to generate chardev_args ?  (improve it to match the new request)
In this way,  param name should be same with option name.
```
1012         options = self.get_supported_options(backend)                           
1013         params = params.copy_from_keys(options)  
```

Reply:
@yanan-fu The method you mentioned is not recommended by other maintainer, and it is indeed not flexible enough for all conditions. I have no strong opinion on this, just follow the way how we create image.

Reply:
@vivianQizhu  In my option,  L1821~1852 are almost duplicate with qdevices.CharDevice. 
For image, there do not have similar function to do this,   but for chardev, it have.

I also think about delete this part (get_supported_options, copy_from_keys, format_params) from CharDevice, but we can since it is useful for hotplug/ hot-unplug. Some other cases use it.

If use the current design, i am afraid we have to change both of these two parts if qemu have change in the future.

Reply:
@yanan-fu All the chardev device should be created from the new introduced wrappers in the further, so there will be no other case use it. chardev usually is a set of devices, it can not be created alone. That's why it is similar with image: a complete device is composed by a set of device(bus/backend/frontend). We need to handle other factors related to environment besides options, like machine types, and the relation between devices, which are not proper to handle in device object, which means, CharDevice is not able to handle all the options. Thus, we have to define a wrapper anyway. 
In conclusion, the options of the there part of a complete device should be handled together, so not in device level.

Reply:
@yanan-fu And, for other device that use chardev as backend(like nic), they should handle the options in its device creation part, not inside CharDevice object either.  So here we take a real device as an entirety in our design, not a single command line.

Reply:
@vivianQizhu  L1835~L1863 have nothing to do with machine type and others, it is only related with chardev options,  and ChaDevice.get_supported_options do the same thing( together with  copy_from_keys).  Correct me if i am wrong.

For this question:  `so there will be no other case use it`:  CharDevice support hotplug/hot-unplug,  `qemu_guest_agent_hotplug` use it.  How can we run hotplug without CharDevice ?  

If no CharDevice, current implement is good for me,  but why don't we call it directly while we can  not delete it.

Reply:
@yanan-fu 
My point is, they should be handled together with all related devices, not in QDevice object level.
For hotplug, it should be all replaced with simple_hotplug().

Reply:
@yanan-fu Well, actually I was about to only use CharDevice at the beginning. @luckyh Would you help to share your thought here? Thanks.